File: src/xuezh/core/process.py
from __future__ import annotations

from dataclasses import dataclass
import shutil
import subprocess
from typing import Sequence


class ToolMissingError(RuntimeError):
    def __init__(self, tool: str) -> None:
        super().__init__(f"Required tool not found on PATH: {tool}")
        self.tool = tool


class ProcessFailedError(RuntimeError):
    def __init__(self, cmd: Sequence[str], returncode: int, stdout: str, stderr: str) -> None:
        super().__init__(f"Process failed with code {returncode}: {' '.join(cmd)}")
        self.cmd = list(cmd)
        self.returncode = returncode
        self.stdout = stdout
        self.stderr = stderr


@dataclass(frozen=True)
class ProcessResult:
    stdout: str
    stderr: str
    returncode: int


def ensure_tool(name: str) -> str:
    path = shutil.which(name)
    if not path:
        raise ToolMissingError(name)
    return path


def run_checked(cmd: Sequence[str]) -> ProcessResult:
    proc = subprocess.run(list(cmd), capture_output=True, text=True)
    if proc.returncode != 0:
        raise ProcessFailedError(cmd, proc.returncode, proc.stdout, proc.stderr)
    return ProcessResult(stdout=proc.stdout, stderr=proc.stderr, returncode=proc.returncode)

File: internal/xuezh/process/process.go
package process

import (
	"errors"
	"fmt"
	"os/exec"
	"strings"
)

type ToolMissingError struct {
	Tool string
}

func (e ToolMissingError) Error() string {
	return fmt.Sprintf("Required tool not found on PATH: %s", e.Tool)
}

type ProcessFailedError struct {
	Cmd        []string
	ReturnCode int
	Stdout     string
	Stderr     string
}

func (e ProcessFailedError) Error() string {
	return fmt.Sprintf("Process failed with code %d: %s", e.ReturnCode, strings.Join(e.Cmd, " "))
}

type ProcessResult struct {
	Stdout     string
	Stderr     string
	ReturnCode int
}

func EnsureTool(name string) (string, error) {
	path, err := exec.LookPath(name)
	if err != nil {
		return "", ToolMissingError{Tool: name}
	}
	return path, nil
}

func RunChecked(cmd []string) (ProcessResult, error) {
	if len(cmd) == 0 {
		return ProcessResult{}, errors.New("empty command")
	}
	c := exec.Command(cmd[0], cmd[1:]...)
	stdout, err := c.Output()
	if err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return ProcessResult{}, ProcessFailedError{Cmd: cmd, ReturnCode: exitErr.ExitCode(), Stdout: string(stdout), Stderr: string(exitErr.Stderr)}
		}
		return ProcessResult{}, err
	}
	return ProcessResult{Stdout: string(stdout), Stderr: "", ReturnCode: 0}, nil
}
